localstack/services/s3/notifications.py

```diff
@@ -0,0 +1,513 @@
+import json
+import logging
+from concurrent.futures import ThreadPoolExecutor
+from typing import Dict, List, Optional, Tuple, TypedDict, Union
+from urllib.parse import quote
+
+from botocore.exceptions import ClientError
+from moto.s3.models import FakeBucket, FakeKey
+
+from localstack.aws.api import RequestContext
+from localstack.aws.api.events import PutEventsRequestEntry
+from localstack.aws.api.s3 import (
+    Event,
+    EventBridgeConfiguration,
+    EventList,
+    LambdaFunctionArn,
+    LambdaFunctionConfiguration,
+    NotificationConfiguration,
+    NotificationConfigurationFilter,
+    NotificationId,
+    QueueArn,
+    QueueConfiguration,
+    TopicArn,
+    TopicConfiguration,
+)
+from localstack.config import DEFAULT_REGION
+from localstack.services.s3.models import get_moto_s3_backend
+from localstack.services.s3.utils import (
+    _create_invalid_argument_exc,
+    get_bucket_from_moto,
+    get_key_from_moto_bucket,
+)
+from localstack.utils.aws import aws_stack
+from localstack.utils.aws.aws_stack import ArnData, parse_arn
+from localstack.utils.strings import short_uid
+from localstack.utils.time import timestamp_millis
+
+LOG = logging.getLogger(__name__)
+
+EVENT_OPERATION_MAP = {
+    "PutObject": Event.s3_ObjectCreated_Put,
+    "CopyObject": Event.s3_ObjectCreated_Copy,
+    "CompleteMultipartUpload": Event.s3_ObjectCreated_CompleteMultipartUpload,
+    "PutObjectTagging": Event.s3_ObjectTagging_Put,
+    "DeleteObjectTagging": Event.s3_ObjectTagging_Delete,
+    "DeleteObject": Event.s3_ObjectRemoved_Delete,
+}
+
+HEADER_AMZN_XRAY = "X-Amzn-Trace-Id"
+
+
+class S3NotificationContent(TypedDict):
+    s3SchemaVersion: str
+    configurationId: NotificationId
+    bucket: Dict[str, str]  # todo
+    object: Dict[str, Union[str, int]]  # todo
+
+
+class EventRecord(TypedDict):
+    eventVersion: str
+    eventSource: str
+    awsRegion: str
+    eventTime: str
+    eventName: str
+    userIdentity: Dict[str, str]
+    requestParameters: Dict[str, str]
+    responseElements: Dict[str, str]
+    s3: S3NotificationContent
+
+
+class Notification(TypedDict):
+    Records: List[EventRecord]
+
+
+class S3EventNotificationContext:
+    def __init__(self, context: RequestContext):
```

Comment:
nit: maybe we could make this a dataclass, and create a special `from_request_context(context: RequestContext) -> S3EventNoticicationContext` factory method. that way the `S3EventNotificationContext` could be created without a request context, plus we have all the fields and their types documented.